File: app/src/main/java/com/jbrunton/mymovies/api/resources/PagedCollection.java
package com.jbrunton.mymovies.api.resources;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public abstract class PagedCollection<T> {
    private Integer page;
    private Integer total_results;
    private Integer total_pages;
    private List<T> results;

    protected <S> List<S> toCollection(Function<T, S> converter) {
        return results.stream()
                .map(converter)
                .collect(Collectors.toList());
    }
}

File: networking/src/main/java/com/jbrunton/networking/resources/PagedCollection.kt
package com.jbrunton.networking.resources

abstract class PagedCollection<T> {
    abstract val page: Int
    abstract val total_results: Int
    abstract val total_pages: Int
    abstract val results: List<T>

    protected fun <S> toCollection(converter: (T) -> S): List<S> {
        return results.map { converter.invoke(it) }
    }
}
